### DISCUSSION TITLE
Automate the CVE Remediation process for Java Third-Party Libraries

### DISCUSSION THREAD

**Discussion by abstractj** [keycloak/keycloak#21498] [Ideas]
Involved: @abstractj @jonkoops @vmuzikar 

## Problem Statement

Currently, we rely on a manual process to update our dependencies whenever there's a Common Vulnerabilities and Exposures (CVE) report. These include reviewing security scanner alerts in the GitHub Security section and updating the dependencies based on Quarkus, both of which can be time-consuming and error-prone.

Monitoring and performing the analysis of security scanner alerts require significant time, and it's crystal clear, the need for automation to reduce the TOIL.

## Requirements

The requirements described here focus on automating CVE remediation and reducing manual steps:

1. Automatic micro-release updates: Set up the automation to upgrade to the next micro release.
2. Targeting: The automation tool should target the `main` and the patch releases branch.
3. Labeling: The labels `area/dependencies` and `status/triage` should be applied, as these improve visibility for our teams.
4. Automatic submission of pull-requests: A new pull-request containing the updates should be submitted automatically.
5. Filtering: The automation tool should have the ability to filter out dependencies which, for certain reasons, we choose not to update.

## Action Plan 

The strategy to implement these changes can happen in two stages:

#### Stage 1

Enable the automation tool to alert us whenever there is a new update available for our dependencies, without automatically submitting a new pull-request. This step allows us to test and refine, tweak the tool configuration without flooding our repository with pull-requests. 

That will require the manual triage of those dependencies to ensure they are relevant. That will also allow us to refine the process without disrupting the development flow.

#### Stage 2

Enable the automation tool to automatically submit pull-requests to our repository whenever updates are available. This step will significantly streamline the update process, reducing manual intervention and our response to new CVEs.

**Comment by abstractj** [comment]
cc @keycloak/maintainers

**Comment by ahus1** [comment]
I support this approach, and I hope the tooling supports the labeling we look for. 

I know that Dependabot can limit the concurrent number of pull requests it creates. If such a tool has an option like this, we safely experiment with a small number of concurrent PRs which are opened.

**Comment by jonkoops** [comment]
Very much a proponent of this, we simply cannot keep up with this manually, and we should not if there is a clear automation (even partly) path.

I imagine that we will start Stage 1 with an allowlist of dependencies that should be upgraded by the tooling, and slowly work towards Stage 2, where we move to a denylist of dependencies we want to hold back. WDYT?

**Reply by abstractj** [reply]
@jonkoops +1 and if we take the Dependabot route. I'm afraid that allowlist is not supported. Still, I don't see any issues about us starting experimenting with it now that Keycloak 22 was released.

**Reply by jonkoops** [reply]
There are alternatives (also open-souce) that we could use. I believe [Renovate](https://docs.renovatebot.com/) does support this use-case for example.

**Comment by stianst** [comment]
@mposolda @vmuzikar @pskopek FIY, also see comments on https://github.com/keycloak/keycloak/issues/24265
